### DISCUSSION TITLE
Error when creating train.spacy

### DISCUSSION THREAD

**Discussion by DarkSoliditi** [explosion/spaCy#9004] [Help: Coding & Implementations] [answered]
I am not sure what I am doing wrong but I am trying to create a train.spacy file. There seems to be an issue with my training data but I don't see a syntax error as it is very similar to the spacy example. The spacy example will create a train.spacy file but mine won't. Any help on this matter is greatly appreciated. Thanks in advance!

I've also tried this code in multiple environments with the same results.

the spacy example:

```
training_data = [
  ("Tokyo Tower is 333m tall.", [(0, 11, "BUILDING")]),
]
```

my training data:

```
training_data = [
    ("For and in consideration.", [(11, 23, "VALUE")]),
]
```

```
nlp = spacy.blank("en")
db = DocBin()

for text, annotations in training_data:
    doc = nlp(text)
    ents = []
    for start, end, label in annotations:
        span = doc.char_span(start, end, label=label)
        ents.append(span)
    doc.ents = ents
    db.add(doc)
db.to_disk("./train.spacy")
```

```
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_321742/2369426394.py in <module>
     16         span = doc.char_span(start, end, label=label)
     17         ents.append(span)
---> 18     doc.ents = ents
     19     db.add(doc)
     20 db.to_disk("./trainb.spacy")

~/train/lib/python3.8/site-packages/spacy/tokens/doc.pyx in spacy.tokens.doc.Doc.ents.__set__()

~/train/lib/python3.8/site-packages/spacy/tokens/doc.pyx in spacy.tokens.doc.get_entity_info()

TypeError: object of type 'NoneType' has no len()
```

**Comment by polm** [comment, marked as answer]
Your character offsets are wrong, so `doc.char_span` returns `None`, which is not a valid ent. You can check it with this code:

```
import spacy
from spacy.tokens import DocBin

training_data = [
    ("For and in consideration.", [(11, 23, "VALUE")]),
]

nlp = spacy.blank("en")
db = DocBin()

for text, annotations in training_data:
    doc = nlp(text)
    ents = []
    for start, end, label in annotations:
        print(doc.text[start:end])
        span = doc.char_span(start, end, label=label)
        ents.append(span)
    print(ents)
    doc.ents = ents
    db.add(doc)
db.to_disk("./train.spacy")
```

Your example data is pulling out the string `consideratio`. Since that doesn't correspond to a token `char_span` just returns `None`. If you look at [the docs](https://spacy.io/api/doc#char_span) you can change how misaligned character offsets are handled, though usually it's best to just fix your offsets.

**Reply by DarkSoliditi** [reply]
Ok yea I see how that works now. That makes a ton of sense. I hadn't gotten to the API documentation yet. Just got over half-way through the documentation on the Usage page. This documentation on Doc.char_span is super helpful! Thanks so much!
